Linearize method for System2.0

RussTedrake opened the pull request:
ready for some comments/discussion @david-german-tri .  include #3967 (which should be merged first).  Will want to squash-merge this one as I had some non-compiling intermediate commits.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4012)
<!-- Reviewable:end -->

david-german-tri commented:
#3967 has merged; this should rebase pretty cleanly.

---

Review status: 0 of 13 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVffE4KUfgwK_dWwVVv:b-4qsegt)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Notes transcribed from f2f.

---

Review status: 0 of 13 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/affine_system.cc, line 130 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfk1sPcigmp1avnRIo:-KVfk1sQxF6h4BYgzZpr:b2ralp7) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L130)):*
> ```cpp
>   autodiff_context->SetTimeStateAndParametersFrom(context);
> 
>   const auto& x0 = context.get_continuous_state_vector().CopyToVector();
> ```

BTW, we could consider adding syntactic sugar for this (though I would rather eliminate the need to copy, see below)

---

*[drake/systems/framework/primitives/affine_system.cc, line 130 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfk6pBPqhfYcfghJTS:-KVfk6pBPqhfYcfghJTT:b-svfuer) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L130)):*
> ```cpp
>   autodiff_context->SetTimeStateAndParametersFrom(context);
> 
>   const auto& x0 = context.get_continuous_state_vector().CopyToVector();
> ```

Consider using `auto` less throughout this file; it's kind of hard to follow what is what type.

---

*[drake/systems/framework/primitives/affine_system.cc, line 135 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfjQWQeXzXhRIGN4MZ:-KVfjQWQeXzXhRIGN4M_:bhqoixl) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L135)):*
> ```cpp
>   Eigen::VectorXd u0 = Eigen::VectorXd::Zero(num_inputs);
>   if (num_inputs>0) {
>     u0 = system.EvalEigenVectorInput(context,0);
> ```

BTW, I need to look carefully at whether making this public interacts badly with #3455.  I think it's OK, but I haven't scrutinized it yet.

---

*[drake/systems/framework/primitives/affine_system.cc, line 143 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfj1n7a9I00Aw6QsOS:-KVfj1n7a9I00Aw6QsOT:bozxgh5) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L143)):*
> ```cpp
> 
>   if (num_inputs>0) {
>     auto input_vector = std::make_unique<BasicVector<AutoDiffXd>>(system.get_input_port(0).get_size());
> ```

Could just use `num_inputs` here

---

*[drake/systems/framework/primitives/affine_system.cc, line 144 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfl0ESdSf8qbOKtc07:-KVfl0ESdSf8qbOKtc08:b-qa8bdf) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L144)):*
> ```cpp
>   if (num_inputs>0) {
>     auto input_vector = std::make_unique<BasicVector<AutoDiffXd>>(system.get_input_port(0).get_size());
>     input_vector->get_mutable_value() = std::get<1>(autodiff_args);
> ```

Can use `SetFromVector` here

---

*[drake/systems/framework/primitives/affine_system.cc, line 145 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfj6t5dJSDBWpZI466:-KVfj6t5dJSDBWpZI467:b-ox2x02) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L145)):*
> ```cpp
>     auto input_vector = std::make_unique<BasicVector<AutoDiffXd>>(system.get_input_port(0).get_size());
>     input_vector->get_mutable_value() = std::get<1>(autodiff_args);
>     autodiff_context->SetInputPort(
> ```

BTW we should have syntactic sugar for this.  `FixInputPort(vector)`, or similar.

---

*[drake/systems/framework/primitives/affine_system.cc, line 153 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfjcLqr5IIrgPv9I9R:-KVfjcLqr5IIrgPv9I9S:b5r6e2l) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L153)):*
> ```cpp
>   auto autodiff_xdot_vec = autodiff_xdot->CopyToVector();
> 
>   auto AB = math::autoDiffToGradientMatrix(autodiff_xdot_vec);
> ```

BTW, we should make versions of these `math::` functions that operate on VectorBase also.  If nothing else, they could do the CopyToVector internally, but we can probably avoid the copy altogether.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVfl89pcUMMLiAHaCKo:b-7ypnv)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 0 of 13 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/affine_system.cc, line 130 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfk6pBPqhfYcfghJTS:-KVfqjFvLulbYzzYujoM:bxn52t) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L130)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Consider using `auto` less throughout this file; it's kind of hard to follow what is what type.
</details>
removed almost all of them.

---

*[drake/systems/framework/primitives/affine_system.cc, line 135 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfjQWQeXzXhRIGN4MZ:-KVfqopjPzeP4frHutJT:b-2dvq8j) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L135)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, I need to look carefully at whether making this public interacts badly with #3455.  I think it's OK, but I haven't scrutinized it yet.
</details>
will definitely wait for your OK.

---

*[drake/systems/framework/primitives/affine_system.cc, line 143 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfj1n7a9I00Aw6QsOS:-KVfrBGNEk5hKIQX8n3V:bb7f2mf) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L143)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Could just use `num_inputs` here
</details>
done

---

*[drake/systems/framework/primitives/affine_system.cc, line 144 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfl0ESdSf8qbOKtc07:-KVfrBvVNfdWIOa38jWE:bb7f2mf) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L144)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Can use `SetFromVector` here
</details>
done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
> #3967 has merged; this should rebase pretty cleanly.

friendly ping re. rebase

---

Review status: 0 of 13 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVftPDeZ9CBN-k9ZcCH:b3j1a87)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Here are a few more comments!  I think this should proceed to merge, with #4021 as a boilerplate reduction pass.  

---

Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/affine_system.cc, line 130 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfk1sPcigmp1avnRIo:-KVfvwJ9ObDS5gPQRChp:b-8semiv) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L130)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, we could consider adding syntactic sugar for this (though I would rather eliminate the need to copy, see below)
</details>
#4021

---

*[drake/systems/framework/primitives/affine_system.cc, line 135 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfjQWQeXzXhRIGN4MZ:-KVfvUTCipPgEssLx3SV:b-1o7icz) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L135)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> will definitely wait for your OK.
</details>
OK, I am convinced this is fine.  Even if the system pointer was extracted from a Diagram, its inputs are hooked up and triggering evaluation from outside will work.  I also can't come up with any situation in which this breaks abstraction; if you can get hold of a System pointer, you are not another System, but actually some user code which should have that level of access.  (As is the case here.)

---

*[drake/systems/framework/primitives/affine_system.cc, line 145 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfj6t5dJSDBWpZI466:-KVfvzeaefjc8XnavAY5:b-8semiv) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L145)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW we should have syntactic sugar for this.  `FixInputPort(vector)`, or similar.
</details>
#4021

---

*[drake/systems/framework/primitives/affine_system.cc, line 153 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfjcLqr5IIrgPv9I9R:-KVfw0RWl_4ke5Jf5GqT:b-8semiv) ([raw file](https://github.com/robotlocomotion/drake/blob/de37731d2d7e95fdbbd9b727d2b89c6013b0158a/drake/systems/framework/primitives/affine_system.cc#L153)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, we should make versions of these `math::` functions that operate on VectorBase also.  If nothing else, they could do the CopyToVector internally, but we can probably avoid the copy altogether.
</details>
#4021

---

*[drake/systems/framework/primitives/affine_system.cc, line 120 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVftn0oMhBFSeSzudEp:-KVftn0oMhBFSeSzudEq:b-7s4sqq) ([raw file](https://github.com/robotlocomotion/drake/blob/d215d597202d6357e570ccb136efe0eddd1849e8/drake/systems/framework/primitives/affine_system.cc#L120)):*
> ```cpp
>   DRAKE_DEMAND(system.get_num_output_ports() <= 1);
> 
>   int num_inputs = (system.get_num_input_ports()>0) ?  system.get_input_port(0).get_size() : 0,
> ```

const

---

*[drake/systems/framework/primitives/affine_system.cc, line 121 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVftsFnkbIWhaOxw58u:-KVftpWh9EClxd5w-o0I:b-7s4sqq) ([raw file](https://github.com/robotlocomotion/drake/blob/d215d597202d6357e570ccb136efe0eddd1849e8/drake/systems/framework/primitives/affine_system.cc#L121)):*
> ```cpp
> 
>   int num_inputs = (system.get_num_input_ports()>0) ?  system.get_input_port(0).get_size() : 0,
>       num_outputs = (system.get_num_output_ports()>0) ? system.get_output_port(0).get_size() : 0;
> ```

const

---

*[drake/systems/framework/primitives/affine_system.cc, line 131 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVftzPuCA0JjUdpP4EZ:-KVftzPuCA0JjUdpP4E_:b-7s4sqq) ([raw file](https://github.com/robotlocomotion/drake/blob/d215d597202d6357e570ccb136efe0eddd1849e8/drake/systems/framework/primitives/affine_system.cc#L131)):*
> ```cpp
> 
>   const Eigen::VectorXd& x0 = context.get_continuous_state_vector().CopyToVector();
>   int num_states = x0.size();
> ```

const

---

*[drake/systems/framework/primitives/affine_system.h, line 97 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfuqMcymcPqLqU0MBH:-KVfuqMcymcPqLqU0MBI:b-8k7bg3) ([raw file](https://github.com/robotlocomotion/drake/blob/d215d597202d6357e570ccb136efe0eddd1849e8/drake/systems/framework/primitives/affine_system.h#L97)):*
> ```objc
> /// even though the result may be affine (not simply linear).
> ///
> /// Note that the context must also have its inputs initialized, e.g. using
> ```

s/initialized/connected

---

*[drake/systems/framework/primitives/affine_system.h, line 97 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfuv3QJHxGhT_NMXbC:-KVfuv3QJHxGhT_NMXbD:buvdjxp) ([raw file](https://github.com/robotlocomotion/drake/blob/d215d597202d6357e570ccb136efe0eddd1849e8/drake/systems/framework/primitives/affine_system.h#L97)):*
> ```objc
> /// even though the result may be affine (not simply linear).
> ///
> /// Note that the context must also have its inputs initialized, e.g. using
> ```

consider: "either from Diagram wiring, or to a FreestandingInputPort: [example code]"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVfw3KAqKh2le1OBTZp:bx0l1yn)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/examples/Pendulum/test/CMakeLists.txt, line 23 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfwE7bQuOwWznEtf_I:-KVfwE7bQuOwWznEtf_J:by0u7w6) ([raw file](https://github.com/robotlocomotion/drake/blob/8aa79201348cc361405b6691340c846e1ae8a16c/drake/examples/Pendulum/test/CMakeLists.txt#L23)):*
> ```CMake
> endif()
> ```

merge hiccup?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVfwS-LCbRRdaQ6lGgV:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/affine_system.h, line 97 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfuv3QJHxGhT_NMXbC:-KVfzucsDWmvC3ZEupbT:brci0cs) ([raw file](https://github.com/robotlocomotion/drake/blob/d215d597202d6357e570ccb136efe0eddd1849e8/drake/systems/framework/primitives/affine_system.h#L97)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> consider: "either from Diagram wiring, or to a FreestandingInputPort: [example code]"
</details>
i'm not sure that actually works.  if i have a diagram with an input connected, then the diagram system no longer has inputs.  so the linearized system would come out with zero inputs.

---

*[drake/systems/framework/primitives/affine_system.h, line 97 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfuqMcymcPqLqU0MBH:-KVg-POUfdb5lZqIYYlb:bcerfin) ([raw file](https://github.com/robotlocomotion/drake/blob/d215d597202d6357e570ccb136efe0eddd1849e8/drake/systems/framework/primitives/affine_system.h#L97)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> s/initialized/connected
</details>
see below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/affine_system.h, line 97 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfuv3QJHxGhT_NMXbC:-KVg00QlAKBQJhAZW51u:bf35zbn) ([raw file](https://github.com/robotlocomotion/drake/blob/d215d597202d6357e570ccb136efe0eddd1849e8/drake/systems/framework/primitives/affine_system.h#L97)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> i'm not sure that actually works.  if i have a diagram with an input connected, then the diagram system no longer has inputs.  so the linearized system would come out with zero inputs.
</details>
Yep, I was unclear.  I meant that you might have `System<T>* MyCustomDiagram<T>::some_particular_subsystem()`, which already has its inputs wired up.   Another cut:

"The inputs in the Context must be connected, either to the output of some upstream System within a Diagram, or to a FreestandingInputPort."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/system.h, line 510 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVgRbrMjPoiSupQULIL:-KVgRbrMjPoiSupQULIM:bsjzw20) ([raw file](https://github.com/robotlocomotion/drake/blob/c958186d3bc71c238a3f419bbd5ee06e43d2ed2b/drake/systems/framework/system.h#L510)):*
> ```objc
>   }
> 
> public:
> ```

from cpplint - +1 indent

---

*[drake/systems/framework/system.h, line 556 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVgReB4VkphRpLLVd6D:-KVgReB4VkphRpLLVd6E:bsjzw20) ([raw file](https://github.com/robotlocomotion/drake/blob/c958186d3bc71c238a3f419bbd5ee06e43d2ed2b/drake/systems/framework/system.h#L556)):*
> ```objc
>   }
> 
> protected:
> ```

from cpplint - +1 indent

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/examples/Pendulum/test/CMakeLists.txt, line 23 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfwE7bQuOwWznEtf_I:-KVgBR-Z-B5myd2Nc3zO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8aa79201348cc361405b6691340c846e1ae8a16c/drake/examples/Pendulum/test/CMakeLists.txt#L23)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> merge hiccup?
</details>
Done.

---

*[drake/systems/framework/system.h, line 510 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVgRbrMjPoiSupQULIL:-KVgVulc3Oojs5pKOYAG:b-3mid29) ([raw file](https://github.com/robotlocomotion/drake/blob/c958186d3bc71c238a3f419bbd5ee06e43d2ed2b/drake/systems/framework/system.h#L510)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> from cpplint - +1 indent
</details>
are you ok with it in this form?  i was going to move these methods above the original protected statement.  just did it this way to make it easier to review.  

---

*[drake/systems/framework/primitives/affine_system.cc, line 120 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVftn0oMhBFSeSzudEp:-KVgBSW84g1rXj0XrDWh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d215d597202d6357e570ccb136efe0eddd1849e8/drake/systems/framework/primitives/affine_system.cc#L120)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> const
</details>
Done.

---

*[drake/systems/framework/primitives/affine_system.cc, line 121 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVftsFnkbIWhaOxw58u:-KVgBURcqx4J8OyUwj9i:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d215d597202d6357e570ccb136efe0eddd1849e8/drake/systems/framework/primitives/affine_system.cc#L121)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> const
</details>
Done.

---

*[drake/systems/framework/primitives/affine_system.cc, line 131 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVftzPuCA0JjUdpP4EZ:-KVgBVUjwfQRRBsZMEnX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d215d597202d6357e570ccb136efe0eddd1849e8/drake/systems/framework/primitives/affine_system.cc#L131)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> const
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
+@ggould-tri for platform review (I believe he's today's reviewer)

---

Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/framework/system.h, line 556 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVgReB4VkphRpLLVd6D:-KVj4vEmTv5pq24yNVL-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c958186d3bc71c238a3f419bbd5ee06e43d2ed2b/drake/systems/framework/system.h#L556)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> from cpplint - +1 indent
</details>
Done.

---

*[drake/systems/framework/primitives/affine_system.h, line 97 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfuv3QJHxGhT_NMXbC:-KVj4wMnyGsv6S7KVQGz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d215d597202d6357e570ccb136efe0eddd1849e8/drake/systems/framework/primitives/affine_system.h#L97)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Yep, I was unclear.  I meant that you might have `System<T>* MyCustomDiagram<T>::some_particular_subsystem()`, which already has its inputs wired up.   Another cut:
> 
> "The inputs in the Context must be connected, either to the output of some upstream System within a Diagram, or to a FreestandingInputPort."
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVj4lxITxH5SSolWYNO:bmbysvd)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
@drake-jenkins-bot retest this please

---

Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVjZ33DfkN-CV1Gaghm:bzdwe9o)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 12 files at r2, 1 of 2 files at r6, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVjZSeAzC6w1Sv4Ee67:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I left some minor nits while waiting for something else to compile.

---

Reviewed 1 of 3 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/Pendulum/test/CMakeLists.txt, line 23 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfwE7bQuOwWznEtf_I:-KVjfa_tQXwY6eSVgoTI:bkqif9g) ([raw file](https://github.com/robotlocomotion/drake/blob/8aa79201348cc361405b6691340c846e1ae8a16c/drake/examples/Pendulum/test/CMakeLists.txt#L23)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> Done.
</details>
BTW. No change? (extra blank line?)

---

*[drake/systems/framework/primitives/affine_system.cc, line 54 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVjgS4AUEmrjGWR-0gu:-KVjgS4AUEmrjGWR-0gv:b8b1p8f) ([raw file](https://github.com/robotlocomotion/drake/blob/40be759aac92986dfadf52371bb04a5dca6be23a/drake/systems/framework/primitives/affine_system.cc#L54)):*
> ```cpp
> }
> 
> // setup equivalent system with a different scalar type
> ```

BTW: nit: cappunc (same for all comments in this file).

---

*[drake/systems/framework/primitives/affine_system.h, line 90 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVjgMu_lhUxNUz3vmew:-KVjgMu_lhUxNUz3vmex:b208b0v) ([raw file](https://github.com/robotlocomotion/drake/blob/40be759aac92986dfadf52371bb04a5dca6be23a/drake/systems/framework/primitives/affine_system.h#L90)):*
> ```objc
> };
> 
> /// Takes the first-order Taylor expansion of a System around an nominal
> ```

BTW: nit: "an" should be "a".

---

*[drake/systems/framework/primitives/test/affine_system_test.cc, line 85 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVjgac6ADZ0ZePjkRk5:-KVjgac6ADZ0ZePjkRk6:bb64e1a) ([raw file](https://github.com/robotlocomotion/drake/blob/40be759aac92986dfadf52371bb04a5dca6be23a/drake/systems/framework/primitives/test/affine_system_test.cc#L85)):*
> ```cpp
> }
> 
> // Test that linearizing and affine system returns an identical affine system
> ```

Missing period.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVjgiVznFjqm_uKxchV:b-5ral98)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
-@ggould-tri +@sammy-tri 

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVka05IC9V1-u7htxLJ:b-d2r02w)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
This is very cool -- AutoDiff rocks! One minor comment below. It would also be nice to include a test case that linearizes something nonlinear! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 12 files at r2, 1 of 2 files at r6, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/framework/primitives/affine_system.cc, line 135 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVl15sRIzDKqnBqEx9s:-KVl15sRIzDKqnBqEx9t:bjjas2y) ([raw file](https://github.com/robotlocomotion/drake/blob/40be759aac92986dfadf52371bb04a5dca6be23a/drake/systems/framework/primitives/affine_system.cc#L135)):*
> ```cpp
>   autodiff_context->SetTimeStateAndParametersFrom(context);
> 
>   const Eigen::VectorXd& x0 =
> ```

This should be a local variable rather than a reference since this is a copy (i.e. remove the `&`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVl2mgsFMBjTosYgySF:b7p7oyf)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
*[drake/systems/framework/primitives/test/affine_system_test.cc, line 85 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVl3h0T0PPJqblcZjCU:-KVl3h0T0PPJqblcZjCV:bwwum2d) ([raw file](https://github.com/robotlocomotion/drake/blob/40be759aac92986dfadf52371bb04a5dca6be23a/drake/systems/framework/primitives/test/affine_system_test.cc#L85)):*
> ```cpp
> }
> 
> // Test that linearizing and affine system returns an identical affine system
> ```

BTW typo and->an.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
> It would also be nice to include a test case that linearizes something nonlinear!
I asked Russ about this f2f, and he said he's planning to linearize the Pendulum in a followup.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVl4HQkzB_XOSdA2F2b:bc25r5h)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (pending remaining issues and commit curation)

---

Reviewed 1 of 12 files at r2, 1 of 2 files at r6, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVlGFysWzShq3sOOIt_:b265gsu)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
David is exactly right.  I started writing one (which would have involved some
Gross copy and paste because i can't include the examples headers in my systems test), but then realized that a) the only differnce would be coverage on the autodiff type -- which doesn't need it and b) i'm about to have that coverage with my lqr PR on the pendulum, etc.  

Will make the final edits and push tonight.  Thanks all.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVlHsbaljnHoAIa2z6i:b2wrb6k)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
apologies for the final round of changes after the review.  i (finally) resolved for myself the right way to make this return a LinearSystem instead of an AffineSystem and the fact that this was actually more correct, since if xdot0!=0 we cannot guarantee that an affine model will capture the local dynamics over a finite time duration.  Also having y0!=0 has to be irrelevant, since the approximation only makes sense in the (y-y0) coordinates.

the code change is small (mostly relocating the code to linear_system.cc, etc) but needs another look, I'm afraid.

---

Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/framework/primitives/affine_system.cc, line 54 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVjgS4AUEmrjGWR-0gu:-KVnuwFL-cH6MxwboNdl:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/40be759aac92986dfadf52371bb04a5dca6be23a/drake/systems/framework/primitives/affine_system.cc#L54)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: nit: cappunc (same for all comments in this file).
</details>
Done.

---

*[drake/systems/framework/primitives/affine_system.cc, line 135 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVl15sRIzDKqnBqEx9s:-KVnv2gM5rzs5aE0uEcq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/40be759aac92986dfadf52371bb04a5dca6be23a/drake/systems/framework/primitives/affine_system.cc#L135)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This should be a local variable rather than a reference since this is a copy (i.e. remove the `&`).
</details>
Done.

---

*[drake/systems/framework/primitives/affine_system.h, line 90 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVjgMu_lhUxNUz3vmew:-KVnv792z04LPYHMriHZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/40be759aac92986dfadf52371bb04a5dca6be23a/drake/systems/framework/primitives/affine_system.h#L90)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: nit: "an" should be "a".
</details>
Done.

---

*[drake/systems/framework/primitives/test/affine_system_test.cc, line 85 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVjgac6ADZ0ZePjkRk5:-KVnvCeOMuZiabNnOJE7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/40be759aac92986dfadf52371bb04a5dca6be23a/drake/systems/framework/primitives/test/affine_system_test.cc#L85)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Missing period.
</details>
Done.

---

*[drake/systems/framework/primitives/test/affine_system_test.cc, line 85 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVl3h0T0PPJqblcZjCU:-KVnvD1AAozlvrK6xtvj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/40be759aac92986dfadf52371bb04a5dca6be23a/drake/systems/framework/primitives/test/affine_system_test.cc#L85)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW typo and->an.
</details>
Done.

---

*[drake/examples/Pendulum/test/CMakeLists.txt, line 23 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVfwE7bQuOwWznEtf_I:-KVnubgdmHAzg0GPAPUr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8aa79201348cc361405b6691340c846e1ae8a16c/drake/examples/Pendulum/test/CMakeLists.txt#L23)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW. No change? (extra blank line?)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVo4TMa3c2m_k3RLzfO:bqihu0o)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
fwiw -- pretty sure the ros build is an unrelated CI flake.

liangfok commented:
Yes, it appears to be another instance of https://github.com/RobotLocomotion/drake/issues/3429. I will thus re-open that issue.

liangfok commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (still pending any commirt curation)

---

Reviewed 6 of 7 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVzKtzlPuagPsPVrYB-:b-a7qaor)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 7 of 7 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/framework/primitives/linear_system.h, line 67 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVzLWWLieubijfdar8l:-KVzLWWMnjcylVtq79A9:b8p93k1) ([raw file](https://github.com/robotlocomotion/drake/blob/6211bfc60f692b972716803e3bc137a5844b3882/drake/systems/framework/primitives/linear_system.h#L67)):*
> ```objc
> ///            std::make_unique<FreestandingInputPort>(default_input));
> ///
> /// Note: The inputs, states, and outputs of the returned system are NOT the
> ```

BTW, it took me a while to work out what this note means, but I have no concrete suggestions to improve it.

---

*[drake/systems/framework/primitives/test/linear_system_test.cc, line 87 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVzKxo2h4C9iHxsACCb:-KVzKxo2h4C9iHxsACCc:boi97h9) ([raw file](https://github.com/robotlocomotion/drake/blob/6211bfc60f692b972716803e3bc137a5844b3882/drake/systems/framework/primitives/test/linear_system_test.cc#L87)):*
> ```cpp
> }
> 
> // Test that linearizing an affine system returns an identical affine system.
> ```

BTW, No longer returns an affine system, right?

---

*[drake/systems/framework/primitives/test/linear_system_test.cc, line 88 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVzL0sqh_7pkBI6EpVk:-KVzL0sqh_7pkBI6EpVl:b9lh33z) ([raw file](https://github.com/robotlocomotion/drake/blob/6211bfc60f692b972716803e3bc137a5844b3882/drake/systems/framework/primitives/test/linear_system_test.cc#L88)):*
> ```cpp
> 
> // Test that linearizing an affine system returns an identical affine system.
> GTEST_TEST(testLinearize, fromAffine) {
> ```

BTW, casing: `TestLinearize, FromAffine`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KVzMZMFKKd2zORSefoT:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
will squash merge.  thanks.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/primitives/test/linear_system_test.cc, line 87 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVzKxo2h4C9iHxsACCb:-KW0Bomufu4OgtFTGmNe:b-cypxrc) ([raw file](https://github.com/robotlocomotion/drake/blob/6211bfc60f692b972716803e3bc137a5844b3882/drake/systems/framework/primitives/test/linear_system_test.cc#L87)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, No longer returns an affine system, right?
</details>
updated.

---

*[drake/systems/framework/primitives/test/linear_system_test.cc, line 88 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-KVzL0sqh_7pkBI6EpVk:-KW0Bt5u6nAzyaK4iTBG:b-qssebl) ([raw file](https://github.com/robotlocomotion/drake/blob/6211bfc60f692b972716803e3bc137a5844b3882/drake/systems/framework/primitives/test/linear_system_test.cc#L88)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> BTW, casing: `TestLinearize, FromAffine`
</details>
fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4012#-:-KW0BvjhmUDdznTe0iYS:bcpuvdj)*
<!-- Sent from Reviewable.io -->
